docs/architecture/adr-047-extend-upgrade-plan.md

```diff
@@ -0,0 +1,206 @@
+    1.  If a `checksum` is provided (in the field or as a query param), the checksum of the result of the `url` MUST equal the provided checksum.
+
+Downloading of an `Artifact` will happen the same way that URLs from `info` are currently downloaded.
+
+### Cosmovisor Updates
+
+If the `upgrade-info.json` file does not contain any `UpgradeInstructions`, existing functionality will be maintained.
+
+We will update Cosmovisor to look for and handle the new `UpgradeInstructions` in `upgrade-info.json`.
+If the `UpgradeInstructions` are provided, we will do the following:
+1.  The `info` field will be ignored.
+1.  The `artifacts` field will be used to identify the artifact to download based on the `platform` that Cosmovisor is running in.
+1.  If a `checksum` is provided (either in the field or as a query param in the `url`), and the downloaded artifact has a different checksum, the upgrade process will be interrupted and Cosmovisor will exit with an error.
+1.  If a `pre_run` command is defined, it will be executed at the same point in the process where the `app pre-upgrade` command would have been executed.
+    It will be executed using the same environment as other commands run by Cosmovisor.
+1.  If a `post_run` command is defined, it will be executed after executing the command that restarts the chain.
+    It will be executed in a background process using the same environment as the other commands.
+    Any output generated by the command will be logged.
+    Once complete, the exit code will be logged.
+
+We will update Cosmovisor to allow JSON of the new `UpgradeInstructions` to be allowed in the existing `info` field.
+We will also allow a url in the `info` field to return such JSON.
+When parsing the `info` field, Cosmovisor will first look for the new `UpgradeInstructions` structure; if not found, Cosmovisor will fallback to existing functionality.
+
+## Consequences
+
+### Backwards Compatibility
+
+Since the only change to existing definitions is the addition of the `upgrade` field to the `Plan` message, and that field is optional, there are no backwards incompatibilities with respects to the proto messages.
+Additionally, current behavior will be maintained when no `UpgradeInstructions` are provided, so there are no backwards incompatibilities with respects to either the upgrade module or Cosmovisor.
+
+### Forwards Compatibility
+
+In order to utilize the `UpgradeInstructions` as part of a software upgrade, both of the following must be true:
+1.  The chain must already be using a sufficiently advanced version of the Cosmos SDK.
+1.  The chain's nodes must be using a sufficiently advanced version of Cosmovisor.
+
+### Positive
+
+1.  The structure for defining artifacts is clearer since it is now defined in the proto instead of in documentation.
+1.  Availability of a pre-run command becomes more obvious.
+1.  A post-run command becomes possible.
+
+### Negative
+
+1.  The `Plan` message becomes larger.
```

Comment:
```suggestion
1.  The `Plan` message becomes larger. However this won't be a problem, because proposals require a bond, and they are pruned.
```

Reply:
Proposals do require a bond, but they also cost gas which is affected by message size. And pruning depends on node setup, so it's not accurate to say definitively that they're pruned.

Reply:
Prunning doesn't depend on the node setup. Inactive or passed proposals are pruned.

Reply:
Updated to this:
```md
1.  The `Plan` message becomes larger. This is negligable because A) the `x/upgrades` module only stores at most one upgrade plan, and B) upgrades are rare enough that the increased gas cost isn't a concern.
```